### PR TITLE
debaml(stream-7c): native-only partial/final SAP + two-leg differential (unrouted)

### PR DESCRIPTION
**Pull request by invakid404**
## Phase 7C — native-only streaming SAP + two-leg differential proof (UNROUTED)

Proves a **single native engine owns every admitted streaming partial and the final** with **no per-prefix BAML fallback**, and gates it with a deterministic two-leg BAML-vs-native loopback differential. Still **UNROUTED** (test-harness API only; nothing reachable from `RunStreamOrchestration` or public routes). Host stays **zero-nanollm / CGO-free**.

### What shipped

**Native-only serving parsers + pure preflight (`internal/debaml/stream_serve.go`)**
- `SupportsNativeStream` / `SupportsNativeFinal` — a **pure, socket-free, input-free** schema-graph preflight (lower + validate + `checkSupported` [+ `checkNoStreamAnnotations`]). Total over the admitted type graph; no FFI, no socket.
- `ParseNativeStreamPartial` / `ParseNativeStreamFinal` — the native-only closures. They **never invoke BAML** (I6): a partial decline is a silent no-emit (the orchestrator skips the tick), a final decline is a **terminal invariant** (`ErrDeBAMLNativeStreamUnsupported`), never a re-parse.

**Eligible-prefix gap closure (`internal/debaml/coerce.go`)**
- A content-free open object — bare `{`, dangling key `{"name"`, dangling colon `{"name":` — now **CLAIMS** BAML parse-stream's all-filler object (`{"name":null,"age":null}`) **byte-exact**, reusing the already-proven missing-field filler logic (the `≥1-present-field` guard is removed for the empty-object case).
- Scoped to an open **object** candidate: bare prose / a just-opened fence with no `{` still decline at *extraction*, so BAML's `allow_as_string→class` recovery is never mis-emitted natively (a safe under-claim). Final parse is untouched.

**Schema-support admission row (`nativeserve/admission`, §5.5)**
- The STREAM claim declines **pre-transport** (I2) any schema the native-only parser cannot own (`ReasonStreamSchemaUnsupported`). The unary lane is byte-for-byte unchanged (shared `admitCore`, stream-only block).

**The gated two-leg loopback differential (`internal/nativebody/nanollmprepare/dynamic`, the core deliverable)**
- Oracle leg: `dynclient.DynamicStream`/`DynamicStreamRaw` **`WithDeBAML(false)`** — BAML-as-served. Candidate leg: **`WithDeBAML(true)` + native parser** — over the same scripted OpenAI SSE server.
- Asserts, over a corpus (person / long-string / list / numbers / default-fillers / reasoning) × **both public modes**: **byte-exact StreamResult cadence** (kind, flattened partial JSON, raw/reasoning accumulation, reset, final), **exactly one provider request + one accepted TCP connection per leg**, same outbound request bytes, and a native-parser **fallback count == 0** — i.e. **no admitted prefix invoked BAML** for partial or final parsing.
- A separate **native transport leg** (`AdmitStreamClaim` + `execute.RunStream`, the 7B exact one-shot lane) runs against the same server: **one accepted socket**, outbound request parity vs BAML, normalized deltas == streamed text, composed through the native-only final parser end-to-end (the literal §6.2 candidate leg).

**Public wire goldens (`cmd/serve/streamwriter_stream7c_golden_test.go`)**
- The real stream writer serializes a native-owned StreamResult sequence into the SAME public **SSE and NDJSON** bytes for **both modes**, without changing the wire format.

### Invariants held
- **I6** one engine owns partial + final — native-only closures, no per-prefix BAML fallback (differential fallback count == 0).
- **I1** flag-off / unrouted identity — the new stream-coerce path runs only under de-BAML; flag-off is byte-identical BAML; the native transport lane stays unrouted.
- **I2/I3/I4** decline is pre-transport; one physical request / one accepted socket per leg (real TCP `ConnState` counting).
- **I8** bounded observability — the schema decline adds one fixed `Stage/Reason` token; no secrets.
- Host + default BAML-only worker remain **zero-nanollm / CGO-free** (`scripts/check-host-zero-nanollm.sh` 4/4).

### Evidence (all green locally)
- `gofmt`; `CGO_ENABLED=0 go build ./...`; `go vet` — clean.
- `go test ./internal/... ./bamlutils/... ./cmd/serve/…` — **3066 + 486** pass (incl. the wire goldens and the updated stream tests).
- `nativeserve` gated (`nanollm_integration`) — admission / execute / parity / canary all pass; unary admission byte-for-byte unchanged.
- `nanollmprepare/dynamic` gated (`integration && nanollm_integration`) — the 6c differential **and** the new 7c two-leg differential pass (both modes, **zero mismatches, zero fallbacks, one socket per leg**).
- `integration` (`-tags integration`, live BAML CFFI) — `TestBamlfuzzParseRecovery` (incl. the flipped `open_brace`/`first_key`/`colon`/`key_only` streaming prefixes, held **byte-exact vs live BAML v0.223**) + `TestDeBAMLStream_*` seam tests pass.
- Native-worker packaging E2E build + tar freshness pass; host-zero-nanollm 4/4.

### Cross-module pin (expected, documented — same class as 7B)
`nativeserve/admission` now imports `internal/debaml.SupportsNativeStream` (new in 7C, not in the `eb48282c` pin). The native-worker **packaging build compiles against local sources → green**; only the external **`nativeserve-goget`** lane goes **red** pre-merge. Resolved at integration by re-pinning the three first-party requires to the 7C master merge commit + regenerating `nativeworker_module.tar` (documented in `nativeserve/go.mod`, NOTE Phase 7C). Do **not** break the packaging build to satisfy goget.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<a href="https://cubic.dev/pr/invakid404/baml-rest/pull/631?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Expanded Phase 7C native-only streaming with schema preflight checks and a terminal error for unsupported stream schemas.
  * Added `DynamicParseRaw` to return flattened worker output without absent-optional injection or schema reordering.
* **Bug Fixes**
  * Improved JSONish comment handling in both final and streaming parsing (including unterminated/dangling comments).
  * Improved numeric/string rendering and certainty (e.g., `inf`/`nan`, `-0`), refined streaming prefix claiming, unclosed-final completion, and string escape decoding (including streaming).
  * Fixed JSONish trailing-comma acceptance in more cases.
* **Tests**
  * Expanded golden, admission, and integration differential/parity coverage for streaming, including fragmentation and edge cases.

### CODE REVIEW

File: internal/debaml/fix.go [hunk 1]
```diff
@@ -324,8 +329,15 @@ func (p *fixer) parseUnquotedScalar(term string) (value, error) {
...
+			// BAML would consume this comma and read greedily — decline. EXCEPTION: a
+			// comma immediately followed by '}' is a TRAILING comma with NO following
+			// field to greedy-swallow, so BAML's greedy read (`36,` then the int/float/
+			// bool coercer trims the trailing comma back to 36) produces the SAME value
+			// as native's clean read of `36` — CLAIM it. (A comma followed by tight
+			// FIELD content is the genuine InObjectValue cascade, which stays declined;
+			// admission also excludes any NON-last unquoted scalar, so only this LAST-
+			// scalar trailing-comma reaches here.)
 			return value{}, errFixUnsupported
 		}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/fix.go:342]: Not actionable on 7C: this guard intentionally matches the live BAML InObjectValue cadence, which recognizes only a space or LF after the comma. A tab or CR is a tight greedy-recovery trigger, not a standard-JSON whitespace path to claim; declining avoids a byte-different emission. The reason is documented beside both final and stream guards.

File: internal/debaml/stream_serve.go [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+	if err != nil {
+		return err
+	}
+	if err := checkSupported(bundle); err != nil {
+		return err
+	}
+	// The single string-absorbing-field root class diverges on the FINAL too
+	// (LIVE-CAPTURED: a bare `"just a string"` final makes BAML error while native
+	// over-emits {"field":"just a string"}), so decline it here as well.
+	return checkStreamRootSupported(bundle)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/stream_serve.go:942]: Not actionable in the 7C native-stream lane. nativeserve admission gates every streaming request with SupportsNativeStream; ParseNativeStreamFinal is the final closure of that same admitted stream, not a broad unary-final ownership API. Keeping its support cut-line no broader than stream admission is deliberate, because widening only SupportsNativeFinal would not admit a real streaming request.

File: internal/debaml/stream_serve.go [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+func ParseNativeStreamPartial(ctx context.Context, s *bamlutils.DynamicOutputSchema, accumulated string) (json.RawMessage, error) {
+	res, err := Parse(ctx, bamlutils.DeBAMLParseRequest{
+		Raw:          accumulated,
+		OutputSchema: s,
+		Stream:       true,
+	})
+	if err != nil {
+		// Decline (sentinel) OR claimed failure: on the partial lane both are a
+		// silent no-emit. NEVER a BAML fallback. The stream continues.
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/stream_serve.go:990]: Not actionable: debaml Parse is a pure CPU parser and does not observe or return context-cancellation errors; orchestration owns cancellation. Therefore this error branch cannot swallow a cancelled-context error. Adding ctx.Err here would change the established partial no-emit contract without fixing an actual cancellation path.

File: internal/debaml/parse.go [hunk 4]
```diff
@@ -169,6 +195,168 @@ func parseStream(req bamlutils.DeBAMLParseRequest) (bamlutils.DeBAMLParseResult,
...
+
+// typeHasScalarMapValue reports whether t contains (directly or nested) a map whose
+// VALUE type is an unquoted scalar — an object-value position that greedy-reads.
+func typeHasScalarMapValue(t schema.Type) bool {
+	switch t.Kind {
+	case schema.TypeMap:
+		if t.Value != nil && (typeIsUnquotedScalar(*t.Value) || typeHasScalarMapValue(*t.Value)) {
+			return true
+		}
+		return t.Value != nil && typeHasScalarMapValue(*t.Value)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/parse.go:275]: No runtime defect: this is a behavior-preserving low-value simplification only. It does not affect the 7C stream boundary or proof, so the final reviewed tip intentionally avoids churn for this nit.

File: internal/debaml/parse_test.go [hunk 5]
```diff
@@ -1163,13 +1163,49 @@ func TestParse_FixingFencedJSONish(t *testing.T) {
...
+		// value) — orthogonal, so those cases are not asserted here.
+		`{"name": "Ada", "age": 36} /* dangling to EOF`,
+		`{"name": "Ada", "age": 36}/*`,
+		`{"name": "Ada", "age": 36} // dangling to EOF`,
+		`{"name": "Ada", "age": 36}//`,
+	} {
+		mustParse(t, s, raw, `{"name":"Ada","age":36}`)
+	}
+	// A comment marker inside a STRING value is not a comment — kept verbatim.
+	mustParse(t, s, `{"name": "a/*b*/c", "age": 36}`, `{"name":"a/*b*/c","age":36}`)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/parse_test.go:1198]: Coverage request only: the implementation already keeps comment markers literal inside a double-quoted string. This additional spelling is test-breadth, not a production defect, and belongs to the owner-accepted residual 7C test-coverage work rather than a new blocker.

File: integration/stream_admitted_acceptance_test.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+		out, e := bamlutils.FlattenDynamicOutput(flat)
+		if e != nil {
+			out = flat
+		}
+		if out, e = bamlutils.InjectAbsentOptionals(out, s); e != nil {
+			return string(flat)
+		}
+		if out, e = bamlutils.ReorderDynamicOutputBySchema(out, s); e != nil {
+			return string(flat)
+		}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply integration/stream_admitted_acceptance_test.go:64]: This is differential-harness hardening, not a runtime parser finding. The fallback preserves the raw public bytes when the normalizer cannot transform a form; it does not make a native parser success. Broader normalizer fail-fast coverage is owner-accepted residual test-breadth work, not a new 7C code defect.

File: integration/stream_typespace_differential_test.go [hunk 7]
```diff
@@ -0,0 +1,779 @@
...
+		case "list":
+			// list bad-child: a non-coercible element for a typed list (BAML drops it).
+			switch f.t.elem.kind {
+			case "int", "litint":
+				inputs = append(inputs, batInput{name: "badchild_" + f.name, raw: rootObj(root, "", map[string]string{f.name: `["1",2,"bad",4]`})})
+			case "enum":
+				inputs = append(inputs, batInput{name: "badchild_" + f.name, raw: rootObj(root, "", map[string]string{f.name: `["RED","NOPE","GREEN"]`})})
+			case "litstr":
+				inputs = append(inputs, batInput{name: "badchild_" + f.name, raw: rootObj(root, "", map[string]string{f.name: `["A","NOPE","A"]`})})
+			}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply integration/stream_typespace_differential_test.go:896]: Correct observation about the generator representation, but this requests an additional differential battery cell, not a native-behavior defect. It belongs to the owner-accepted residual test-breadth work after the count-pinned 289-shape milestone, so it is tracked rather than a fresh 7C blocker.

File: internal/nativebody/nanollmprepare/dynamic/stream_differential_integration_test.go [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+	w.Header().Set("Content-Type", "text/event-stream")
+	w.WriteHeader(http.StatusOK)
+	flusher.Flush()
+
+	write := func(payload string) {
+		// A small inter-frame pause lets the in-process consumer drain each partial
+		// before the next frame, so the orchestrator's NONBLOCKING partial send never
+		// drops under load — making the cadence deterministic (the drop is best-effort
+		// and timing-dependent, not a semantic difference).
+		time.Sleep(2 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/nativebody/nanollmprepare/dynamic/stream_differential_integration_test.go]: @coderabbitai Fixed in `becbe751`. Removed the `time.Sleep(2 * time.Millisecond)` timing hack — delivery is now deterministic by construction, not by scheduler luck. The orchestrator's nonblocking partial send (`trySendPartialShared`, `default:` drop) targets the dynclient result channel, which the `Baml_Rest_Dynamic` adapter allocates as `make(chan …, 100)`; every scripted fixture emits far fewer than 100 partial-producing deltas, so the `default:` drop branch can never fire regardless of timing. Replaced the sleep with a comment documenting that guarantee, and re-ran `TestStreamDifferential -count=3/-count=5` — stable, no dropped partials.

File: internal/debaml/extract.go [hunk 9]
```diff
@@ -4,6 +4,78 @@ import (
...
+				i++
+			}
+			// Leave the newline (if any) for the next iteration to preserve line
+			// structure; the comment body itself collapses to one space.
+			b.WriteByte(' ')
+			continue
+		}
+		b.WriteByte(c)
+		i++
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/extract.go:99]: @coderabbitai Fixed in `becbe751`. `stripJSONComments` now tracks BOTH `"` and `'` as string delimiters (active `strQuote`), so a `/*…*/` or `//…` marker inside a single-quoted value is content, not a stripped comment. Measured byte-exact vs live BAML v0.223: `{'name':'a/*b*/c','age':5}` → `{"name":"a/*b*/c",…}` and `{'name':'a//b',…}` → `{"name":"a//b",…}` (native previously corrupted the first to `a c`). Crucially, a single quote opens a string ONLY in a structural value/key START position (start, or after `{ [ , :`) — matching BAML, which does NOT treat a bare prose apostrophe as a delimiter: `x's /* c */ {…}` still strips the comment (verified BAML strips it too; naive symmetric tracking would have wrongly protected it). Regression added to `TestParse_Comments`. The 289-shape differential + deferred byte-check stayed byte-identical (0 divergences, 0 over-claim).

File: internal/debaml/complete_final.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+			out = append(out, s[start:i]...)
+			if len(fr) > 0 {
+				top().expectValue = false
+			}
+			if i >= len(s) {
+				trailTokStart = tokOut // token ran to EOF: may be a partial number
+			} else if len(fr) > 0 {
+				commit()
+			}
+		}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/complete_final.go:150]: @coderabbitai Fixed in `becbe751`. The EOF completion state machine now (a) opens/closes SINGLE-quoted strings (`strQuote`, merged `case '"', '\''`) — a `'` reaches that switch only at a fresh token start (a `'` mid-bareword is consumed by the token loop), so no extra structural gating is needed; and (b) recognises JSONish UNQUOTED keys — an unquoted token in key position ends at `:` and is tracked like a closed key, so a valueless key tail drops to the class default. Measured byte-exact vs live BAML: `{'name':'Ad` → `{"name":"Ad",…defaults}`; `{a:` / `{a` / `{'a'` → both-default; `{a:["x"]` → `{"a":["x"],…}`; `{name:` / `{'name':` → both error (missing required) — identical to BAML on each. New EOF-prefix parity fixtures (single-q keys/values/elements + unquoted keys) added to `TestStream7CUnclosedFinalParity`: +125 prefixes, 0 mismatches (394 total). Admitted matrix + differential unchanged (289/134/155, 0 divergences).

File: internal/debaml/complete_final.go [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+		return s, false
+	}
+	switch {
+	case inStr:
+		if strIsKey {
+			out = out[:top().safeLen] // drop the partial key
+		} else {
+			out = append(out, '"') // recover the trailing value/element string
+			commit()
+		}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/complete_final.go:171]: @coderabbitai Fixed in `becbe751`. At EOF inside an unterminated string, if the last byte was a lone `\` (esc set), the completion now emits the escape pair (`\\`) BEFORE the closing quote, so the recovered JSON is valid and the backslash stays literal. Measured byte-exact vs live BAML: `{"name":"Ad\` → `{"name":"Ad\\",…}` (content `Ad\`). (A backslash inside a SINGLE-quoted string is the long-standing #583 fixer deferral — `parseSingleQuoted` declines any backslash — so that case stays a pure under-approximation, never an over-claim; unchanged.) Regression added to `TestCompleteUnclosedFinal` plus the end-to-end prefix parity. Differential stayed byte-identical.

File: internal/debaml/stream_serve.go [hunk 12]
```diff
@@ -0,0 +1,964 @@
...
+		}
+		return false
+	}
+	for i := range root.Fields {
+		if checkNested(root.Fields[i].Type) {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/stream_serve.go]: @coderabbitai Confirmed and fixed in `2d2da364`.

A self- or mutually-recursive class graph (`Node{next:Node}`, `A{b:B}`/`B{a:A}`) did stack-overflow `SupportsNativeStream` — reproduced empirically (`fatal error: stack overflow`). The other admission gates iterate `b.Classes` flatly (cycle-safe) and `classSubtreeHasList` already carries a `seen` guard; `checkNested` was the only unguarded recursive descent, exactly as flagged.

Fix (two parts):
1. A new `typeGraphHasClassCycle` gate declines a cyclic class graph pre-transport — a proper 3-colour DFS (a back-edge to a class still on the DFS stack is a cycle; re-reaching a fully-explored class via a second acyclic path — a DAG diamond — is NOT). Cyclic schemas are outside the finite acyclic admitted core, so declining is the conservative disposition (an under-claim, never an over-claim).
2. `checkNested` now threads a `seen` map (matching its sibling `classSubtreeHasList`) as defense-in-depth for any direct caller.

Verified: `Node{next:Node}` and `A<->B` now DECLINE cleanly (no crash); a DAG diamond (`root{x:Leaf, y:Leaf}`) is NOT mistaken for a cycle and stays admitted. Regression: `TestSupportsNativeStream_CyclicClassGraphDeclines`. The admitted matrix is unchanged (the generative differential enumerates only bounded acyclic shapes): 289 shapes (134 admitted / 155 declined), 0 divergences, 0 admission mismatches — byte-identical to before.

File: integration/stream_unclosed_final_test.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+		return norm(s, res.Data), true
+	}
+	// The PRODUCTION native-only FINAL closure (Parse + completeUnclosedFinal on decline).
+	nat := func(s *bamlutils.DynamicOutputSchema, raw string) (string, bool) {
+		j, e := debaml.ParseNativeStreamFinal(context.Background(), s, raw)
+		if e != nil {
+			return "", false
+		}
+		return norm(s, j), true
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply integration/stream_unclosed_final_test.go:70]: Not a correctness issue on the rebase. These closures remain intentionally test-local: the admitted-matrix harness supports both streaming and final DynamicParse modes, while this EOF-final proof needs one preserve=true non-stream BAML oracle paired with the native final closure. Extracting a generic testutil API would broaden scope and couple unrelated test setup without changing the byte-exact proof. Keeping the focused harness local avoids that churn.
